who_l3_smart_tools/core/terminology/terminology.py

```diff
@@ -0,0 +1,256 @@
+        self._process_row_required_fields(kwargs)
+        self._process_value_set()
+
+    def _process_row_required_fields(self, extra_args: dict):
+        """
+        Processes the required fields of the row.
+
+        Args:
+            extra_args (dict): Additional arguments.
+
+        Returns:
+            dict: The processed row data.
+        """
+        converted_row = {
+            "id": self.row.pop(self.schema.id),
+            "name": self.row.pop(self.schema.name),
+            "datatype": self.row.pop(self.schema.datatype),
+            "description": self.row.pop(self.schema.description),
+            "concept_class": self.concept_class,
+        }
+        for index, name in enumerate(self.schema.additional_names):
+            converted_row[f"name[{index+1}]"] = self.row.pop(name)
+        for index, description in enumerate(self.schema.additional_descriptions):
+            converted_row[f"description[{index+1}]"] = self.row.pop(description)
+        converted_row.update(extra_args)
+        self.converted_row = converted_row
+
+    def _process_value_set(self):
+        if self.converted_row.get("datatype").lower() == "codes":
+            if not self.value_set:
+                raise ValueError(f"Value set is required for datatype 'codes'. {self.converted_row}")
+            codings = self.value_set["type"]["coding"]
+            codings.append(
+                {
+                    "system": "http://hl7.org/fhir/v2/0203",
+                    "code": self.converted_row["id"].replace(".", ""),
+                    "display": self.converted_row["name"],
+                }
+            )
+        elif self.converted_row.get("datatype").lower() == "coding":
+            self.value_set = {
+                "system": "https://fhir.staging.aws.openconceptlab.org",
+                "type": {"coding": [], "text": self.converted_row["name"]},
+                "value": self.converted_row["id"].replace(".", ""),
+            }
+        else:
+            self.value_set = None
+
+    def process_for_csv(self):
+        """
+        Processes the row for CSV output.
+        """
+        for key, value in self.row.items():
+            if key in self.schema.exclude_columns:
+                continue
+            if self.schema.include_columns and key not in self.schema.include_columns:
+                continue
+            self.converted_row[self.schema.format_extras_for_csv(key)] = value
+
+    def process_for_json(self):
+        """
+        Processes the row for JSON output.
+        """
+        self.converted_row["extras"] = {}
+        for key, value in self.row.items():
+            if key in self.schema.exclude_columns:
+                continue
+            if self.schema.include_columns and key not in self.schema.include_columns:
+                continue
+            self.converted_row["extras"][
+                self.schema.format_extras_for_json(key)
+            ] = value
+
+
+class ConceptTerminology:
```

Comment:
Could we rename this? It's not clear to me what this represents? Does it generate a OCL Source or OCL Collection? Does it collect the data from the data dictionary? "ConceptTerminology" sounds like nonsense to my ears.

Reply:
It generates [concepts](https://docs.openconceptlab.org/en/latest/oclapi/apireference/concepts.html). This is actually the core of what the entire PR is doing. I was under the impression that the main output for this task is to convert concepts from the input xls into an output similar to [this](https://docs.openconceptlab.org/en/latest/oclapi/apireference/bulkimporting.html#concepts). Do you have suggestions on what is more appropriate to call it?

Reply:
Maybe go with `ConceptResourceGenerator`... longer but more accurate. Yes, you're right about the task; I just think the "Terminology" word is throwing me, since it has a domain-specific meaning (a terminology is a set of concepts, what this generates is resources for creating a representation of a terminology or something along those lines).